File: app/src/main/java/com/gmail/jorgegilcavazos/ballislife/features/boxscore/BoxScorePresenter.java
package com.gmail.jorgegilcavazos.ballislife.features.boxscore;

import com.gmail.jorgegilcavazos.ballislife.features.model.BoxScoreValues;
import com.gmail.jorgegilcavazos.ballislife.network.API.NbaGamesService;
import com.gmail.jorgegilcavazos.ballislife.util.schedulers.BaseSchedulerProvider;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableSingleObserver;

public class BoxScorePresenter {

    private BoxScoreView view;
    private NbaGamesService nbaService;
    private BaseSchedulerProvider schedulerProvider;
    private CompositeDisposable disposables;

    public BoxScorePresenter(BoxScoreView view, NbaGamesService nbaService,
                             BaseSchedulerProvider schedulerProvider) {
        this.view = view;
        this.nbaService = nbaService;
        this.schedulerProvider = schedulerProvider;
    }

    public void start() {
        disposables = new CompositeDisposable();
    }

    public void loadBoxScore(String gameId, final int teamSelected) {
        view.setLoadingIndicator(true);
        view.hideBoxScore();
        view.hideLoadMessage();
        disposables.clear();
        disposables.add(nbaService.boxScore(gameId)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribeWith(new DisposableSingleObserver<BoxScoreValues>() {
                    @Override
                    public void onSuccess(BoxScoreValues boxScoreValues) {
                        if (isViewAttached()) {
                            if (boxScoreValues.getHls().getPstsg() != null) {
                                if (teamSelected == BoxScoreFragment.LOAD_AWAY) {
                                    view.showVisitorBoxScore(boxScoreValues);
                                } else {
                                    view.showHomeBoxScore(boxScoreValues);
                                }
                            } else {
                                view.showBoxScoreNotAvailableMessage();
                            }
                            view.setLoadingIndicator(false);
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        if (isViewAttached()) {
                            view.showLoadingBoxScoreErrorMessage();
                            view.setLoadingIndicator(false);
                        }
                    }
                })
        );
    }

    public void stop() {
        view = null;
        if (disposables != null) {
            disposables.clear();
        }
    }

    private boolean isViewAttached() {
        return view != null;
    }
}

File: app/src/main/java/com/gmail/jorgegilcavazos/ballislife/features/boxscore/BoxScorePresenter.kt
package com.gmail.jorgegilcavazos.ballislife.features.boxscore

import com.gmail.jorgegilcavazos.ballislife.base.BasePresenter
import com.gmail.jorgegilcavazos.ballislife.data.repository.boxscore.BoxScoreRepository
import com.gmail.jorgegilcavazos.ballislife.util.ErrorHandler
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import javax.inject.Inject

class BoxScorePresenter @Inject constructor(
		private val boxScoreRepository: BoxScoreRepository,
		private val disposable: CompositeDisposable,
		private val errorHandler: ErrorHandler): BasePresenter<BoxScoreView>() {

	override fun detachView() {
		disposable.clear()
		super.detachView()
	}

	fun loadBoxScore(gameId: String, selectedTeam: BoxScoreSelectedTeam, forceNetwork: Boolean) {
		disposable.clear()
		boxScoreRepository.boxScore(gameId, forceNetwork)
				.subscribe({ boxModel: BoxScoreUIModel ->
					if (boxModel.inProgress) {
						view.hideBoxScore()
						view.setLoadingIndicator(true)
						view.showBoxScoreNotAvailableMessage(false)
					}

					if (boxModel.notAvailable) {
						view.setLoadingIndicator(false)
						view.showBoxScoreNotAvailableMessage(true)
						view.hideBoxScore()
					}

					if (boxModel.success) {
						when (selectedTeam) {
							BoxScoreSelectedTeam.HOME -> view.showHomeBoxScore(boxModel.boxScore!!.game)
							BoxScoreSelectedTeam.VISITOR -> view.showVisitorBoxScore(boxModel.boxScore!!.game)
						}

						view.setLoadingIndicator(false)
						view.showBoxScoreNotAvailableMessage(false)
					}
				}, { e: Throwable ->
					view.showUnknownErrorToast(errorHandler.handleError(e))
				})
				.addTo(disposable)
	}
}
